File: ScoreForge/Scoring/ScoreCalculator.cs
using ScoreForge.Models;
using ScoreForge.Templates;

namespace ScoreForge.Scoring;

public sealed record PlayerStanding(
    Guid PlayerId,
    string PlayerName,
    int Total,
    int Rank,
    bool IsLeader,
    bool IsWinner);

public sealed record GameSnapshot(
    IReadOnlyList<PlayerStanding> Standings,
    int CurrentRound,
    bool IsComplete,
    string? WinnerName);

public static class ScoreCalculator
{
    public static GameSnapshot Calculate(Game game, IGameTemplate template)
    {
        var totals = game.Players.ToDictionary(
            p => p.Id,
            p => game.Events.Where(e => e.PlayerId == p.Id).Sum(e => e.Points));

        var currentRound = game.Events
            .Where(e => e.RoundNumber.HasValue)
            .Select(e => e.RoundNumber!.Value)
            .DefaultIfEmpty(0)
            .Max();

        var ordered = OrderPlayers(game.Players, totals, template.WinCondition);
        var isComplete = game.Status == GameStatus.Completed
            || DetectCompletion(game, template, totals, currentRound);

        var winnerIds = isComplete
            ? GetWinnerIds(ordered, totals, template, game)
            : [];

        var standings = ordered.Select((player, index) =>
        {
            var total = totals[player.Id];
            var isWinner = winnerIds.Contains(player.Id);
            var isLeader = index == 0 && ordered.Count > 0
                && totals[ordered[0].Id] == total;
            return new PlayerStanding(
                player.Id,
                player.Name,
                total,
                index + 1,
                isLeader && !isComplete,
                isWinner);
        }).ToList();

        var winnerName = standings.FirstOrDefault(s => s.IsWinner)?.PlayerName;

        return new GameSnapshot(standings, currentRound, isComplete, winnerName);
    }

    public static int NextRoundNumber(Game game)
    {
        var max = game.Events
            .Where(e => e.RoundNumber.HasValue)
            .Select(e => e.RoundNumber!.Value)
            .DefaultIfEmpty(0)
            .Max();
        return max + 1;
    }

    private static List<Player> OrderPlayers(
        IEnumerable<Player> players,
        IReadOnlyDictionary<Guid, int> totals,
        WinCondition winCondition)
    {
        var list = players.ToList();
        return winCondition switch
        {
            WinCondition.LowestTotal => list
                .OrderBy(p => totals[p.Id])
                .ThenBy(p => p.Name, StringComparer.OrdinalIgnoreCase)
                .ToList(),
            _ => list
                .OrderByDescending(p => totals[p.Id])
                .ThenBy(p => p.Name, StringComparer.OrdinalIgnoreCase)
                .ToList()
        };
    }

    private static bool DetectCompletion(
        Game game,
        IGameTemplate template,
        IReadOnlyDictionary<Guid, int> totals,
        int currentRound)
    {
        if (template.WinCondition == WinCondition.FirstToTarget)
        {
            var target = game.TargetScore ?? template.DefaultTargetScore;
            if (target is int t && totals.Values.Any(v => v >= t))
                return true;
        }

        var maxRounds = game.MaxRounds ?? template.DefaultMaxRounds;
        if (maxRounds is int rounds && currentRound >= rounds)
            return true;

        return false;
    }

    private static HashSet<Guid> GetWinnerIds(
        IReadOnlyList<Player> ordered,
        IReadOnlyDictionary<Guid, int> totals,
        IGameTemplate template,
        Game game)
    {
        if (ordered.Count == 0)
            return [];

        if (template.WinCondition == WinCondition.FirstToTarget)
        {
            var target = game.TargetScore ?? template.DefaultTargetScore ?? int.MaxValue;
            var reached = ordered.Where(p => totals[p.Id] >= target).ToList();
            if (reached.Count == 0)
                return [];

            var best = reached.Max(p => totals[p.Id]);
            return reached.Where(p => totals[p.Id] == best).Select(p => p.Id).ToHashSet();
        }

        var winningTotal = totals[ordered[0].Id];
        return ordered.Where(p => totals[p.Id] == winningTotal).Select(p => p.Id).ToHashSet();
    }
}

File: mobile/src/domain/scoreCalculator.ts
import type { Game, Player } from './models';
import type { GameTemplate, WinCondition } from './templates';

export interface PlayerStanding {
  playerId: string;
  playerName: string;
  total: number;
  rank: number;
  isLeader: boolean;
  isWinner: boolean;
}

export interface GameSnapshot {
  standings: PlayerStanding[];
  currentRound: number;
  isComplete: boolean;
  winnerName: string | null;
}

export function calculate(game: Game, template: GameTemplate): GameSnapshot {
  const totals = new Map<string, number>();
  for (const player of game.players) {
    totals.set(
      player.id,
      game.events.filter((e) => e.playerId === player.id).reduce((sum, e) => sum + e.points, 0),
    );
  }

  const currentRound = game.events
    .map((e) => e.roundNumber ?? 0)
    .reduce((max, n) => Math.max(max, n), 0);

  const ordered = orderPlayers(game.players, totals, template.winCondition);
  const isComplete =
    game.status === 'Completed' || detectCompletion(game, template, totals, currentRound);

  const winnerIds = isComplete ? getWinnerIds(ordered, totals, template, game) : new Set<string>();

  const standings: PlayerStanding[] = ordered.map((player, index) => {
    const total = totals.get(player.id) ?? 0;
    const isWinner = winnerIds.has(player.id);
    const isLeader =
      index === 0 && ordered.length > 0 && (totals.get(ordered[0].id) ?? 0) === total;
    return {
      playerId: player.id,
      playerName: player.name,
      total,
      rank: index + 1,
      isLeader: isLeader && !isComplete,
      isWinner,
    };
  });

  return {
    standings,
    currentRound,
    isComplete,
    winnerName: standings.find((s) => s.isWinner)?.playerName ?? null,
  };
}

export function nextRoundNumber(game: Game): number {
  const max = game.events
    .map((e) => e.roundNumber ?? 0)
    .reduce((m, n) => Math.max(m, n), 0);
  return max + 1;
}

function orderPlayers(
  players: Player[],
  totals: Map<string, number>,
  winCondition: WinCondition,
): Player[] {
  const list = [...players];
  if (winCondition === 'LowestTotal') {
    return list.sort(
      (a, b) =>
        (totals.get(a.id) ?? 0) - (totals.get(b.id) ?? 0) ||
        a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }),
    );
  }
  return list.sort(
    (a, b) =>
      (totals.get(b.id) ?? 0) - (totals.get(a.id) ?? 0) ||
      a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }),
  );
}

function detectCompletion(
  game: Game,
  template: GameTemplate,
  totals: Map<string, number>,
  currentRound: number,
): boolean {
  if (template.winCondition === 'FirstToTarget') {
    const target = game.targetScore ?? template.defaultTargetScore;
    if (target != null && [...totals.values()].some((v) => v >= target)) return true;
  }

  const maxRounds = game.maxRounds ?? template.defaultMaxRounds;
  if (maxRounds != null && currentRound >= maxRounds) return true;

  return false;
}

function getWinnerIds(
  ordered: Player[],
  totals: Map<string, number>,
  template: GameTemplate,
  game: Game,
): Set<string> {
  if (ordered.length === 0) return new Set();

  if (template.winCondition === 'FirstToTarget') {
    const target = game.targetScore ?? template.defaultTargetScore ?? Number.MAX_SAFE_INTEGER;
    const reached = ordered.filter((p) => (totals.get(p.id) ?? 0) >= target);
    if (reached.length === 0) return new Set();
    const best = Math.max(...reached.map((p) => totals.get(p.id) ?? 0));
    return new Set(reached.filter((p) => (totals.get(p.id) ?? 0) === best).map((p) => p.id));
  }

  const winningTotal = totals.get(ordered[0].id) ?? 0;
  return new Set(ordered.filter((p) => (totals.get(p.id) ?? 0) === winningTotal).map((p) => p.id));
}
